config/src/keys.rs

```diff
@@ -99,22 +45,66 @@ where
 impl<T> KeyPair<T>
 where
     T: PrivateKey + Serialize + ValidKeyStringExt,
-    T::PublicKeyMaterial: DeserializeOwned + 'static + Serialize + ValidKeyStringExt,
 {
+    /// This transforms a private key into a keypair data structure.
     pub fn load(private_key: T) -> Self {
         let public_key = private_key.public_key();
         Self {
-            private_key: PrivateKeyContainer::Present(private_key),
+            private_key: Some(private_key),
             public_key,
         }
     }
 
-    pub fn public(&self) -> &T::PublicKeyMaterial {
-        &self.public_key
-    }
-
-    /// Beware, this destroys the private key from this NodeConfig
+    /// Takes the key from the data structure, calling this function a second time will return None.
     pub fn take_private(&mut self) -> Option<T> {
         self.private_key.take()
     }
+
+    /// Returns the public key part. This always work, even after the private key was taken.
+    pub fn public_key(&self) -> T::PublicKeyMaterial {
+        self.public_key.clone()
+    }
+}
+
+//
+// Serialization and Deserialization functions
+//
+
+/// Serialization for a KeyPair only serializes the private key part (if present).
+impl<T> Serialize for KeyPair<T>
+where
+    T: PrivateKey + Serialize + ValidKeyStringExt,
+{
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+        T: Serialize + ValidKeyStringExt,
+    {
+        match &self.private_key {
+            Some(key) => key.serialize(serializer),
+            None => serializer.serialize_str(""),
+        }
+    }
+}
+
+/// Deserializing a keypair only deserializes the private key, and dynamically derives the public key.
+impl<'de, T> Deserialize<'de> for KeyPair<T>
+where
+    T: PrivateKey + ValidKeyStringExt,
+{
+    fn deserialize<D>(deserializer: D) -> std::result::Result<KeyPair<T>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        match T::deserialize(deserializer) {
+            Ok(private_key) => {
+                let public_key = private_key.public_key();
+                Ok(KeyPair {
+                    private_key: Some(private_key),
+                    public_key,
+                })
```

Comment:
I was under the impression that public_key would be populated upon take_key()

Reply:
but I guess not...

Reply:
I did this originally, but this required KeyPair to be mutable when handled... the current approach is actually much cooler!